File: solutions/C++/clone_graph_bfs.cpp
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node(int _val = 0) { val = _val; }
};

class Solution {
public:
    Node* cloneGraphBFS(Node* node) {
        if (!node) return nullptr;
        
        unordered_map<int, Node*> visited;
        queue<Node*> q;
        q.push(node);
        visited[node->val] = new Node(node->val);
        
        while (!q.empty()) {
            Node* curr = q.front();
            q.pop();
            
            for (Node* neighbor : curr->neighbors) {
                if (!visited.count(neighbor->val)) {
                    visited[neighbor->val] = new Node(neighbor->val);
                    q.push(neighbor);
                }
                visited[curr->val]->neighbors.push_back(visited[neighbor->val]);
            }
        }
        
        return visited[node->val];
    }
};

File: solutions/Rust/clone_graph_bfs.rs
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::{HashMap, VecDeque};

#[derive(Debug)]
pub struct Node {
    pub val: i32,
    pub neighbors: Vec<Rc<RefCell<Node>>>,
}

impl Node {
    pub fn new(val: i32) -> Self {
        Node {
            val,
            neighbors: Vec::new(),
        }
    }
}

fn cloneGraphBFS(node: Option<Rc<RefCell<Node>>>) -> Option<Rc<RefCell<Node>>> {
    match node {
        None => None,
        Some(n) => {
            let mut visited = HashMap::new();
            let mut queue = VecDeque::new();
            
            let val = n.borrow().val;
            let cloned = Rc::new(RefCell::new(Node::new(val)));
            visited.insert(val, Rc::clone(&cloned));
            queue.push_back(n);
            
            while let Some(curr) = queue.pop_front() {
                let neighbors = curr.borrow().neighbors.iter().map(|n| Rc::clone(n)).collect::<Vec<_>>();
                
                for neighbor in neighbors {
                    let nval = neighbor.borrow().val;
                    if !visited.contains_key(&nval) {
                        let cloned_neighbor = Rc::new(RefCell::new(Node::new(nval)));
                        visited.insert(nval, Rc::clone(&cloned_neighbor));
                        queue.push_back(neighbor);
                    }
                    visited.get(&curr.borrow().val).unwrap().borrow_mut().neighbors.push(Rc::clone(visited.get(&nval).unwrap()));
                }
            }
            
            Some(visited.get(&n.borrow().val).unwrap().clone())
        }
    }
}
